File: web_version/pmugg_files/optimizer.js
ms.optimizer = function(nodeStats) {
	this.nodeStats = nodeStats;
	this.prevCost = this.computeCost();
};

ms.optimizer.detailedCost = false;
ms.optimizer.costScale = 0.01;

ms.optimizer.prototype.computeCost = function() {
	var cost = {};

	// Compute line distance from the change.
	var prevLogLineDistances = this.prevCost ? this.prevCost.logLineDistances : 0;	
	cost.logLineDistances = prevLogLineDistances + this.nodeStats.costChange.lineDistance;
	cost.reject = this.nodeStats.costChange.reject || 0;
	var numLines = this.nodeStats.count.line;
	var desiredLines = ms.globalSettings.get('Desired Lines');
	var desiredCost  = ms.globalSettings.get('Desired Lines Cost');
	cost.desiredLines = desiredCost * Math.max((desiredLines - numLines) / desiredLines, -1);
	
	var desiredVertexWeight  = ms.globalSettings.get('Desired Vertex Weight');
	var desirabilitySum = 0;
	this.nodeStats.getElements('vertex').forEach(function(vertex) {
		var decoration = vertex.getState().getType().getDecoration();
		if (decoration) {
			desirabilitySum -= decoration.get('Desirability');
		}
	});
	cost.desiredVertex = desiredVertexWeight * desirabilitySum;
	
	cost.sum = ms.optimizer.costScale *	(cost.logLineDistances + cost.reject + cost.desiredLines + cost.desiredVertex);
	return cost;
};

ms.optimizer.prototype.accept = function(cost) {
	this.prevCost = cost;
};

// Accept or reject the mutation based on the Metropolis algorithm.
ms.optimizer.prototype.isAccepted = function(cost) {
	var difference = this.prevCost.sum - cost.sum;
	var prob = Math.exp(ms.globalSettings.get('Beta') * difference);
	if (Math.random() < prob) {
		this.prevCost = cost;
		return true;
	} else {
		return false;
	}
};

ms.optimizer.prototype.getPrevCost = function() {
	return this.prevCost;
};

// Verify that the cost is restored to its previous value when the model is restored.
ms.optimizer.prototype.verifyCost = function() {
	var costChange = this.computeCost().sum - this.prevCost.sum;
	if (Math.abs(costChange) > 1e-10) {
		ms.alert('Model was not restored correctly.');
		debugger;
	}
};

File: cpp_version/optimizer.cpp
#include "pch.h"
#include "optimizer.h"
#ifdef _CONSOLE
#include "minmax.h"
#endif
#include "settings.h"
#include "graph_drawing/vertex.h"
#include "primitives/vertex_type.h"
#include "util/util.h"
#include <stdexcept>

namespace ms {

bool Optimizer::detailedCost = false;
constexpr double Optimizer::costScale;

Optimizer::Optimizer(const NodeStats& nodeStats)
    : nodeStats(nodeStats) {
    prevCost = computeCost();
}

Optimizer::Cost Optimizer::computeCost() {
    Cost cost;
    cost.logLineDistances =
        prevCost.logLineDistances + nodeStats.getCostChange().lineDistance;
    cost.reject = nodeStats.getCostChange().reject;

    int numLines = nodeStats.getLineCount();
    double desiredLines = globalSettings["Desired Lines"].get<double>();
    double desiredCost = globalSettings["Desired Lines Cost"].get<double>();
    cost.desiredLines =
        desiredCost * max((desiredLines - numLines) / desiredLines, -1.0);

    double desiredVertexWeight = globalSettings["Desired Vertex Weight"].get<double>();
    double desirabilitySum = 0.0;
    for (auto* vertex : nodeStats.getVertices()) {
        desirabilitySum -= vertex->getType()->desirability;
    }
    cost.desiredVertex = desiredVertexWeight * desirabilitySum;

    cost.sum = costScale * (cost.logLineDistances + cost.reject +
                            cost.desiredLines + cost.desiredVertex);
    return cost;
}

void Optimizer::accept(const Cost& cost) {
    prevCost = cost;
}

bool Optimizer::isAccepted(const Cost& cost) {
    double beta = globalSettings["Beta"].get<double>();
    // Short-circuit at Beta=0: acceptance probability is exp(0)=1 so the
    // draw is meaningless. Skipping `randomValue()` here keeps the RNG
    // state lock-step with the pre-optimizer always-accept path, so older
    // grammars that haven't been tuned for the optimizer stay bit-identical.
    if (beta == 0.0) {
        prevCost = cost;
        return true;
    }
    double difference = prevCost.sum - cost.sum;
    double prob = std::exp(beta * difference);
    if (randomValue() < prob) {
        prevCost = cost;
        return true;
    }
    return false;
}

const Optimizer::Cost& Optimizer::getPrevCost() const {
    return prevCost;
}

void Optimizer::verifyCost() {
    double costChange = computeCost().sum - prevCost.sum;
    if (std::abs(costChange) > 1e-10) {
        throw std::runtime_error("Model was not restored correctly.");
    }
}

} // namespace ms 